// arb-ts/dist/lib/abi/factories/ArbAggregator__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArbAggregator__factory = void 0;
var ethers_1 = require("ethers");
var ArbAggregator__factory = /** @class */ (function () {
    function ArbAggregator__factory() {
    }
    ArbAggregator__factory.connect = function (address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    };
    return ArbAggregator__factory;
}());
exports.ArbAggregator__factory = ArbAggregator__factory;
var _abi = [
    {
        inputs: [],
        name: 'getDefaultAggregator',
        outputs: [
            {
                internalType: 'address',
                name: '',
                type: 'address',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'address',
                name: 'addr',
                type: 'address',
            },
        ],
        name: 'getPreferredAggregator',
        outputs: [
            {
                internalType: 'address',
                name: '',
                type: 'address',
            },
            {
                internalType: 'bool',
                name: '',
                type: 'bool',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'address',
                name: 'newDefault',
                type: 'address',
            },
        ],
        name: 'setDefaultAggregator',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'address',
                name: 'prefAgg',
                type: 'address',
            },
        ],
        name: 'setPreferredAggregator',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
];
