// arb-ts/dist/lib/abi/factories/ArbBLS__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArbBLS__factory = void 0;
var ethers_1 = require("ethers");
var ArbBLS__factory = /** @class */ (function () {
    function ArbBLS__factory() {
    }
    ArbBLS__factory.connect = function (address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    };
    return ArbBLS__factory;
}());
exports.ArbBLS__factory = ArbBLS__factory;
var _abi = [
    {
        inputs: [
            {
                internalType: 'address',
                name: 'addr',
                type: 'address',
            },
        ],
        name: 'getPublicKey',
        outputs: [
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256',
            },
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256',
            },
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256',
            },
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'uint256',
                name: 'x0',
                type: 'uint256',
            },
            {
                internalType: 'uint256',
                name: 'x1',
                type: 'uint256',
            },
            {
                internalType: 'uint256',
                name: 'y0',
                type: 'uint256',
            },
            {
                internalType: 'uint256',
                name: 'y1',
                type: 'uint256',
            },
        ],
        name: 'register',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
];
