// arb-ts/dist/lib/abi/factories/ArbFunctionTable__factory 2.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArbFunctionTable__factory = void 0;
var ethers_1 = require("ethers");
var ArbFunctionTable__factory = /** @class */ (function () {
    function ArbFunctionTable__factory() {
    }
    ArbFunctionTable__factory.connect = function (address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    };
    return ArbFunctionTable__factory;
}());
exports.ArbFunctionTable__factory = ArbFunctionTable__factory;
var _abi = [
    {
        inputs: [
            {
                internalType: 'address',
                name: 'addr',
                type: 'address',
            },
            {
                internalType: 'uint256',
                name: 'index',
                type: 'uint256',
            },
        ],
        name: 'get',
        outputs: [
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256',
            },
            {
                internalType: 'bool',
                name: '',
                type: 'bool',
            },
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'address',
                name: 'addr',
                type: 'address',
            },
        ],
        name: 'size',
        outputs: [
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'bytes',
                name: 'buf',
                type: 'bytes',
            },
        ],
        name: 'upload',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
];
