// arb-ts/dist/lib/abi/factories/ArbosTest__factory 2.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArbosTest__factory = void 0;
var ethers_1 = require("ethers");
var ArbosTest__factory = /** @class */ (function () {
    function ArbosTest__factory() {
    }
    ArbosTest__factory.connect = function (address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    };
    return ArbosTest__factory;
}());
exports.ArbosTest__factory = ArbosTest__factory;
var _abi = [
    {
        inputs: [
            {
                internalType: 'address',
                name: 'addr',
                type: 'address',
            },
        ],
        name: 'getAccountInfo',
        outputs: [],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'address',
                name: 'addr',
                type: 'address',
            },
        ],
        name: 'getMarshalledStorage',
        outputs: [],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'address',
                name: 'addr',
                type: 'address',
            },
            {
                internalType: 'bool',
                name: 'isEOA',
                type: 'bool',
            },
            {
                internalType: 'uint256',
                name: 'balance',
                type: 'uint256',
            },
            {
                internalType: 'uint256',
                name: 'nonce',
                type: 'uint256',
            },
            {
                internalType: 'bytes',
                name: 'code',
                type: 'bytes',
            },
            {
                internalType: 'bytes',
                name: 'initStorage',
                type: 'bytes',
            },
        ],
        name: 'installAccount',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
];
