// arb-ts/dist/lib/abi/factories/ArbOwner__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArbOwner__factory = void 0;
var ethers_1 = require("ethers");
var ArbOwner__factory = /** @class */ (function () {
    function ArbOwner__factory() {
    }
    ArbOwner__factory.connect = function (address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    };
    return ArbOwner__factory;
}());
exports.ArbOwner__factory = ArbOwner__factory;
var _abi = [
    {
        inputs: [],
        name: 'addToReserveFunds',
        outputs: [],
        stateMutability: 'payable',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'address',
                name: 'addr',
                type: 'address',
            },
            {
                internalType: 'uint256',
                name: 'pluggableId',
                type: 'uint256',
            },
        ],
        name: 'bindAddressToPluggable',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'bytes',
                name: 'marshalledCode',
                type: 'bytes',
            },
        ],
        name: 'continueCodeUpload',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [],
        name: 'finishCodeUploadAsArbosUpgrade',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'uint256',
                name: 'id',
                type: 'uint256',
            },
            {
                internalType: 'bool',
                name: 'keepState',
                type: 'bool',
            },
        ],
        name: 'finishCodeUploadAsPluggable',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [],
        name: 'getFeeRecipients',
        outputs: [
            {
                internalType: 'address',
                name: '',
                type: 'address',
            },
            {
                internalType: 'address',
                name: '',
                type: 'address',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'address',
                name: 'newOwnerAddr',
                type: 'address',
            },
        ],
        name: 'giveOwnership',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'address',
                name: 'addr',
                type: 'address',
            },
        ],
        name: 'setFairGasPriceSender',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'address',
                name: 'netFeeRecipient',
                type: 'address',
            },
            {
                internalType: 'address',
                name: 'congestionFeeRecipient',
                type: 'address',
            },
        ],
        name: 'setFeeRecipients',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'bool',
                name: 'enabled',
                type: 'bool',
            },
        ],
        name: 'setFeesEnabled',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'uint256',
                name: 'speedLimitPerBlock',
                type: 'uint256',
            },
            {
                internalType: 'uint256',
                name: 'gasPoolMax',
                type: 'uint256',
            },
            {
                internalType: 'uint256',
                name: 'maxTxGasLimit',
                type: 'uint256',
            },
        ],
        name: 'setGasAccountingParams',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'uint256',
                name: 'blocksPerSend',
                type: 'uint256',
            },
        ],
        name: 'setSecondsPerSend',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [],
        name: 'startCodeUpload',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
];
