// arb-ts/dist/lib/abi/factories/ArbStatistics__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArbStatistics__factory = void 0;
var ethers_1 = require("ethers");
var ArbStatistics__factory = /** @class */ (function () {
    function ArbStatistics__factory() {
    }
    ArbStatistics__factory.connect = function (address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    };
    return ArbStatistics__factory;
}());
exports.ArbStatistics__factory = ArbStatistics__factory;
var _abi = [
    {
        inputs: [],
        name: 'getStats',
        outputs: [
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256',
            },
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256',
            },
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256',
            },
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256',
            },
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256',
            },
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
];
