// arb-ts/dist/lib/abi/factories/Cloneable__factory 2.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Cloneable__factory = void 0;
var contracts_1 = require("@ethersproject/contracts");
var Cloneable__factory = /** @class */ (function (_super) {
    __extends(Cloneable__factory, _super);
    function Cloneable__factory(signer) {
        return _super.call(this, _abi, _bytecode, signer) || this;
    }
    Cloneable__factory.prototype.deploy = function (overrides) {
        return _super.prototype.deploy.call(this, overrides || {});
    };
    Cloneable__factory.prototype.getDeployTransaction = function (overrides) {
        return _super.prototype.getDeployTransaction.call(this, overrides || {});
    };
    Cloneable__factory.prototype.attach = function (address) {
        return _super.prototype.attach.call(this, address);
    };
    Cloneable__factory.prototype.connect = function (signer) {
        return _super.prototype.connect.call(this, signer);
    };
    Cloneable__factory.connect = function (address, signerOrProvider) {
        return new contracts_1.Contract(address, _abi, signerOrProvider);
    };
    return Cloneable__factory;
}(contracts_1.ContractFactory));
exports.Cloneable__factory = Cloneable__factory;
var _abi = [
    {
        inputs: [],
        stateMutability: 'nonpayable',
        type: 'constructor',
    },
    {
        inputs: [],
        name: 'isMaster',
        outputs: [
            {
                internalType: 'bool',
                name: '',
                type: 'bool',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
];
var _bytecode = '0x6080604052348015600f57600080fd5b506000805460ff1916600117905560868061002b6000396000f3fe6080604052348015600f57600080fd5b506004361060285760003560e01c80636f791d2914602d575b600080fd5b60336047565b604080519115158252519081900360200190f35b60005460ff169056fea264697066735822122058b9ee7254c959b6503b4633dc6b8c4f765d5ee87fdf5bbcc9e718528d29f83a64736f6c634300060b0033';
