// arb-ts/dist/lib/abi/factories/ERC165Upgradeable__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ERC165Upgradeable__factory = void 0;
var ethers_1 = require("ethers");
var ERC165Upgradeable__factory = /** @class */ (function () {
    function ERC165Upgradeable__factory() {
    }
    ERC165Upgradeable__factory.connect = function (address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    };
    return ERC165Upgradeable__factory;
}());
exports.ERC165Upgradeable__factory = ERC165Upgradeable__factory;
var _abi = [
    {
        inputs: [
            {
                internalType: 'bytes4',
                name: 'interfaceId',
                type: 'bytes4',
            },
        ],
        name: 'supportsInterface',
        outputs: [
            {
                internalType: 'bool',
                name: '',
                type: 'bool',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
];
