// arb-ts/dist/lib/abi/factories/ExtCodeSizeTest__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExtCodeSizeTest__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        stateMutability: 'nonpayable',
        type: 'constructor',
    },
];
const _bytecode = '0x6080604052348015600f57600080fd5b50601c602860201b60201c565b602457600080fd5b603f565b6000803090506000813b9050600081149250505090565b603f80604c6000396000f3fe6080604052600080fdfea26469706673582212209b402e711d4609798f6713ca1688a0dec3bc36ccc6e5618818852fe445760e9c64736f6c634300060b0033';
class ExtCodeSizeTest__factory extends ethers_1.ContractFactory {
    constructor(signer) {
        super(_abi, _bytecode, signer);
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    attach(address) {
        return super.attach(address);
    }
    connect(signer) {
        return super.connect(signer);
    }
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.ExtCodeSizeTest__factory = ExtCodeSizeTest__factory;
ExtCodeSizeTest__factory.bytecode = _bytecode;
ExtCodeSizeTest__factory.abi = _abi;
