// arb-ts/dist/lib/abi/factories/IArbCustomToken__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IArbCustomToken__factory = void 0;
var ethers_1 = require("ethers");
var IArbCustomToken__factory = /** @class */ (function () {
    function IArbCustomToken__factory() {
    }
    IArbCustomToken__factory.connect = function (address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    };
    return IArbCustomToken__factory;
}());
exports.IArbCustomToken__factory = IArbCustomToken__factory;
var _abi = [
    {
        inputs: [
            {
                internalType: 'address',
                name: 'account',
                type: 'address',
            },
            {
                internalType: 'uint256',
                name: 'amount',
                type: 'uint256',
            },
        ],
        name: 'bridgeMint',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
];
