// arb-ts/dist/lib/abi/factories/IArbTokenBridge__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IArbTokenBridge__factory = void 0;
var ethers_1 = require("ethers");
var IArbTokenBridge__factory = /** @class */ (function () {
    function IArbTokenBridge__factory() {
    }
    IArbTokenBridge__factory.connect = function (address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    };
    return IArbTokenBridge__factory;
}());
exports.IArbTokenBridge__factory = IArbTokenBridge__factory;
var _abi = [
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: 'address',
                name: 'l1Address',
                type: 'address',
            },
            {
                indexed: true,
                internalType: 'address',
                name: 'l2Address',
                type: 'address',
            },
        ],
        name: 'CustomTokenRegistered',
        type: 'event',
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: 'bool',
                name: 'success',
                type: 'bool',
            },
            {
                indexed: true,
                internalType: 'address',
                name: 'sender',
                type: 'address',
            },
            {
                indexed: true,
                internalType: 'address',
                name: 'dest',
                type: 'address',
            },
            {
                indexed: false,
                internalType: 'uint256',
                name: 'amount',
                type: 'uint256',
            },
            {
                indexed: false,
                internalType: 'bytes',
                name: 'callHookData',
                type: 'bytes',
            },
        ],
        name: 'MintAndCallTriggered',
        type: 'event',
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: 'address',
                name: 'l1Address',
                type: 'address',
            },
            {
                indexed: true,
                internalType: 'address',
                name: 'l2Address',
                type: 'address',
            },
        ],
        name: 'TokenCreated',
        type: 'event',
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: 'address',
                name: 'l1Address',
                type: 'address',
            },
            {
                indexed: true,
                internalType: 'address',
                name: 'account',
                type: 'address',
            },
            {
                indexed: false,
                internalType: 'uint256',
                name: 'amount',
                type: 'uint256',
            },
        ],
        name: 'TokenMigrated',
        type: 'event',
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: 'address',
                name: 'l1Address',
                type: 'address',
            },
            {
                indexed: true,
                internalType: 'address',
                name: 'l2Address',
                type: 'address',
            },
            {
                indexed: true,
                internalType: 'address',
                name: 'sender',
                type: 'address',
            },
            {
                indexed: true,
                internalType: 'address',
                name: 'dest',
                type: 'address',
            },
            {
                indexed: false,
                internalType: 'uint256',
                name: 'amount',
                type: 'uint256',
            },
            {
                indexed: false,
                internalType: 'bool',
                name: 'usedCallHook',
                type: 'bool',
            },
        ],
        name: 'TokenMinted',
        type: 'event',
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: 'uint256',
                name: 'withdrawalId',
                type: 'uint256',
            },
            {
                indexed: true,
                internalType: 'address',
                name: 'l1Address',
                type: 'address',
            },
            {
                indexed: false,
                internalType: 'uint256',
                name: 'amount',
                type: 'uint256',
            },
            {
                indexed: true,
                internalType: 'address',
                name: 'destination',
                type: 'address',
            },
            {
                indexed: true,
                internalType: 'uint256',
                name: 'exitNum',
                type: 'uint256',
            },
        ],
        name: 'WithdrawToken',
        type: 'event',
    },
    {
        inputs: [
            {
                internalType: 'address',
                name: 'l1ERC20',
                type: 'address',
            },
        ],
        name: 'calculateL2TokenAddress',
        outputs: [
            {
                internalType: 'address',
                name: '',
                type: 'address',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'address',
                name: 'l1Address',
                type: 'address',
            },
            {
                internalType: 'address',
                name: 'l2Address',
                type: 'address',
            },
        ],
        name: 'customTokenRegistered',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'address',
                name: 'l1ERC20',
                type: 'address',
            },
            {
                internalType: 'address',
                name: 'sender',
                type: 'address',
            },
            {
                internalType: 'address',
                name: 'destination',
                type: 'address',
            },
            {
                internalType: 'uint256',
                name: 'amount',
                type: 'uint256',
            },
        ],
        name: 'migrate',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'address',
                name: 'l1ERC20',
                type: 'address',
            },
            {
                internalType: 'address',
                name: 'sender',
                type: 'address',
            },
            {
                internalType: 'address',
                name: 'dest',
                type: 'address',
            },
            {
                internalType: 'uint256',
                name: 'amount',
                type: 'uint256',
            },
            {
                internalType: 'bytes',
                name: 'deployData',
                type: 'bytes',
            },
            {
                internalType: 'bytes',
                name: 'callHookData',
                type: 'bytes',
            },
        ],
        name: 'mintFromL1',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'address',
                name: 'l1ERC20',
                type: 'address',
            },
            {
                internalType: 'address',
                name: 'sender',
                type: 'address',
            },
            {
                internalType: 'address',
                name: 'destination',
                type: 'address',
            },
            {
                internalType: 'uint256',
                name: 'amount',
                type: 'uint256',
            },
        ],
        name: 'withdraw',
        outputs: [
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256',
            },
        ],
        stateMutability: 'nonpayable',
        type: 'function',
    },
];
