// arb-ts/dist/lib/abi/factories/IChallengeFactory__factory 2.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IChallengeFactory__factory = void 0;
var ethers_1 = require("ethers");
var IChallengeFactory__factory = /** @class */ (function () {
    function IChallengeFactory__factory() {
    }
    IChallengeFactory__factory.connect = function (address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    };
    return IChallengeFactory__factory;
}());
exports.IChallengeFactory__factory = IChallengeFactory__factory;
var _abi = [
    {
        inputs: [
            {
                internalType: 'address',
                name: '_resultReceiver',
                type: 'address',
            },
            {
                internalType: 'bytes32',
                name: '_executionHash',
                type: 'bytes32',
            },
            {
                internalType: 'uint256',
                name: '_maxMessageCount',
                type: 'uint256',
            },
            {
                internalType: 'address',
                name: '_asserter',
                type: 'address',
            },
            {
                internalType: 'address',
                name: '_challenger',
                type: 'address',
            },
            {
                internalType: 'uint256',
                name: '_asserterTimeLeft',
                type: 'uint256',
            },
            {
                internalType: 'uint256',
                name: '_challengerTimeLeft',
                type: 'uint256',
            },
            {
                internalType: 'contract IBridge',
                name: '_bridge',
                type: 'address',
            },
        ],
        name: 'createChallenge',
        outputs: [
            {
                internalType: 'address',
                name: '',
                type: 'address',
            },
        ],
        stateMutability: 'nonpayable',
        type: 'function',
    },
];
