// arb-ts/dist/lib/abi/factories/IERC1363Receiver__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IERC1363Receiver__factory = void 0;
var ethers_1 = require("ethers");
var IERC1363Receiver__factory = /** @class */ (function () {
    function IERC1363Receiver__factory() {
    }
    IERC1363Receiver__factory.connect = function (address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    };
    return IERC1363Receiver__factory;
}());
exports.IERC1363Receiver__factory = IERC1363Receiver__factory;
var _abi = [
    {
        inputs: [
            {
                internalType: 'address',
                name: 'operator',
                type: 'address',
            },
            {
                internalType: 'address',
                name: 'sender',
                type: 'address',
            },
            {
                internalType: 'uint256',
                name: 'amount',
                type: 'uint256',
            },
            {
                internalType: 'bytes',
                name: 'data',
                type: 'bytes',
            },
        ],
        name: 'onTransferReceived',
        outputs: [
            {
                internalType: 'bytes4',
                name: '',
                type: 'bytes4',
            },
        ],
        stateMutability: 'nonpayable',
        type: 'function',
    },
];
