// arb-ts/dist/lib/abi/factories/IMessageProvider__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IMessageProvider__factory = void 0;
var ethers_1 = require("ethers");
var IMessageProvider__factory = /** @class */ (function () {
    function IMessageProvider__factory() {
    }
    IMessageProvider__factory.connect = function (address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    };
    return IMessageProvider__factory;
}());
exports.IMessageProvider__factory = IMessageProvider__factory;
var _abi = [
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: 'uint256',
                name: 'messageNum',
                type: 'uint256',
            },
            {
                indexed: false,
                internalType: 'bytes',
                name: 'data',
                type: 'bytes',
            },
        ],
        name: 'InboxMessageDelivered',
        type: 'event',
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: 'uint256',
                name: 'messageNum',
                type: 'uint256',
            },
        ],
        name: 'InboxMessageDeliveredFromOrigin',
        type: 'event',
    },
];
