// arb-ts/dist/lib/abi/factories/OneStepProofCommon__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OneStepProofCommon__factory = void 0;
var ethers_1 = require("ethers");
var OneStepProofCommon__factory = /** @class */ (function () {
    function OneStepProofCommon__factory() {
    }
    OneStepProofCommon__factory.connect = function (address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    };
    return OneStepProofCommon__factory;
}());
exports.OneStepProofCommon__factory = OneStepProofCommon__factory;
var _abi = [
    {
        inputs: [
            {
                internalType: 'contract IBridge',
                name: 'bridge',
                type: 'address',
            },
            {
                internalType: 'uint256',
                name: 'initialMessagesRead',
                type: 'uint256',
            },
            {
                internalType: 'bytes32[2]',
                name: 'accs',
                type: 'bytes32[2]',
            },
            {
                internalType: 'bytes',
                name: 'proof',
                type: 'bytes',
            },
            {
                internalType: 'bytes',
                name: 'bproof',
                type: 'bytes',
            },
        ],
        name: 'executeStep',
        outputs: [
            {
                internalType: 'uint64',
                name: 'gas',
                type: 'uint64',
            },
            {
                internalType: 'uint256',
                name: 'totalMessagesRead',
                type: 'uint256',
            },
            {
                internalType: 'bytes32[4]',
                name: 'fields',
                type: 'bytes32[4]',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'contract IBridge',
                name: 'bridge',
                type: 'address',
            },
            {
                internalType: 'uint256',
                name: 'initialMessagesRead',
                type: 'uint256',
            },
            {
                internalType: 'bytes32[2]',
                name: 'accs',
                type: 'bytes32[2]',
            },
            {
                internalType: 'bytes',
                name: 'proof',
                type: 'bytes',
            },
            {
                internalType: 'bytes',
                name: 'bproof',
                type: 'bytes',
            },
        ],
        name: 'executeStepDebug',
        outputs: [
            {
                internalType: 'string',
                name: 'startMachine',
                type: 'string',
            },
            {
                internalType: 'string',
                name: 'afterMachine',
                type: 'string',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
];
