// arb-ts/dist/lib/abi/factories/TesterERC20TokenNoMetadata__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TesterERC20TokenNoMetadata__factory = void 0;
var contracts_1 = require("@ethersproject/contracts");
var TesterERC20TokenNoMetadata__factory = /** @class */ (function (_super) {
    __extends(TesterERC20TokenNoMetadata__factory, _super);
    function TesterERC20TokenNoMetadata__factory(signer) {
        return _super.call(this, _abi, _bytecode, signer) || this;
    }
    TesterERC20TokenNoMetadata__factory.prototype.deploy = function (overrides) {
        return _super.prototype.deploy.call(this, overrides || {});
    };
    TesterERC20TokenNoMetadata__factory.prototype.getDeployTransaction = function (overrides) {
        return _super.prototype.getDeployTransaction.call(this, overrides || {});
    };
    TesterERC20TokenNoMetadata__factory.prototype.attach = function (address) {
        return _super.prototype.attach.call(this, address);
    };
    TesterERC20TokenNoMetadata__factory.prototype.connect = function (signer) {
        return _super.prototype.connect.call(this, signer);
    };
    TesterERC20TokenNoMetadata__factory.connect = function (address, signerOrProvider) {
        return new contracts_1.Contract(address, _abi, signerOrProvider);
    };
    return TesterERC20TokenNoMetadata__factory;
}(contracts_1.ContractFactory));
exports.TesterERC20TokenNoMetadata__factory = TesterERC20TokenNoMetadata__factory;
var _abi = [
    {
        inputs: [],
        stateMutability: 'nonpayable',
        type: 'constructor',
    },
];
var _bytecode = '0x6080604052348015600f57600080fd5b50603f80601d6000396000f3fe6080604052600080fdfea26469706673582212209997ccdb3e94e790e4d6cdaaa8c5e09441c6be420deab63e592e540ce69254ca64736f6c634300060b0033';
