// arb-ts/dist/lib/abi/factories/TokenAddressHandler__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenAddressHandler__factory = void 0;
var ethers_1 = require("ethers");
var TokenAddressHandler__factory = /** @class */ (function () {
    function TokenAddressHandler__factory() {
    }
    TokenAddressHandler__factory.connect = function (address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    };
    return TokenAddressHandler__factory;
}());
exports.TokenAddressHandler__factory = TokenAddressHandler__factory;
var _abi = [
    {
        inputs: [
            {
                internalType: 'address',
                name: '',
                type: 'address',
            },
        ],
        name: 'customL2Token',
        outputs: [
            {
                internalType: 'address',
                name: '',
                type: 'address',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'address',
                name: 'l1Token',
                type: 'address',
            },
        ],
        name: 'isCustomToken',
        outputs: [
            {
                internalType: 'bool',
                name: '',
                type: 'bool',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
];
