// arb-ts/dist/lib/abi/factories/ISequencerInbox__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ISequencerInbox__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: 'uint256',
                name: 'firstMessageNum',
                type: 'uint256',
            },
            {
                indexed: true,
                internalType: 'bytes32',
                name: 'beforeAcc',
                type: 'bytes32',
            },
            {
                indexed: false,
                internalType: 'uint256',
                name: 'newMessageCount',
                type: 'uint256',
            },
            {
                indexed: false,
                internalType: 'uint256',
                name: 'totalDelayedMessagesRead',
                type: 'uint256',
            },
            {
                indexed: false,
                internalType: 'bytes32[2]',
                name: 'afterAccAndDelayed',
                type: 'bytes32[2]',
            },
            {
                indexed: false,
                internalType: 'uint256',
                name: 'seqBatchIndex',
                type: 'uint256',
            },
        ],
        name: 'DelayedInboxForced',
        type: 'event',
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: 'address',
                name: 'addr',
                type: 'address',
            },
            {
                indexed: false,
                internalType: 'bool',
                name: 'isSequencer',
                type: 'bool',
            },
        ],
        name: 'IsSequencerUpdated',
        type: 'event',
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: 'uint256',
                name: 'newMaxDelayBlocks',
                type: 'uint256',
            },
            {
                indexed: false,
                internalType: 'uint256',
                name: 'newMaxDelaySeconds',
                type: 'uint256',
            },
        ],
        name: 'MaxDelayUpdated',
        type: 'event',
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: 'uint256',
                name: 'firstMessageNum',
                type: 'uint256',
            },
            {
                indexed: true,
                internalType: 'bytes32',
                name: 'beforeAcc',
                type: 'bytes32',
            },
            {
                indexed: false,
                internalType: 'uint256',
                name: 'newMessageCount',
                type: 'uint256',
            },
            {
                indexed: false,
                internalType: 'bytes32',
                name: 'afterAcc',
                type: 'bytes32',
            },
            {
                indexed: false,
                internalType: 'bytes',
                name: 'transactions',
                type: 'bytes',
            },
            {
                indexed: false,
                internalType: 'uint256[]',
                name: 'lengths',
                type: 'uint256[]',
            },
            {
                indexed: false,
                internalType: 'uint256[]',
                name: 'sectionsMetadata',
                type: 'uint256[]',
            },
            {
                indexed: false,
                internalType: 'uint256',
                name: 'seqBatchIndex',
                type: 'uint256',
            },
            {
                indexed: false,
                internalType: 'address',
                name: 'sequencer',
                type: 'address',
            },
        ],
        name: 'SequencerBatchDelivered',
        type: 'event',
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: 'uint256',
                name: 'firstMessageNum',
                type: 'uint256',
            },
            {
                indexed: true,
                internalType: 'bytes32',
                name: 'beforeAcc',
                type: 'bytes32',
            },
            {
                indexed: false,
                internalType: 'uint256',
                name: 'newMessageCount',
                type: 'uint256',
            },
            {
                indexed: false,
                internalType: 'bytes32',
                name: 'afterAcc',
                type: 'bytes32',
            },
            {
                indexed: false,
                internalType: 'uint256',
                name: 'seqBatchIndex',
                type: 'uint256',
            },
        ],
        name: 'SequencerBatchDeliveredFromOrigin',
        type: 'event',
    },
    {
        inputs: [],
        name: 'getInboxAccsLength',
        outputs: [
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'uint256',
                name: 'index',
                type: 'uint256',
            },
        ],
        name: 'inboxAccs',
        outputs: [
            {
                internalType: 'bytes32',
                name: '',
                type: 'bytes32',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'address',
                name: 'seq',
                type: 'address',
            },
        ],
        name: 'isSequencer',
        outputs: [
            {
                internalType: 'bool',
                name: '',
                type: 'bool',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'maxDelayBlocks',
        outputs: [
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'maxDelaySeconds',
        outputs: [
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'messageCount',
        outputs: [
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'bytes',
                name: 'proof',
                type: 'bytes',
            },
            {
                internalType: 'uint256',
                name: 'inboxCount',
                type: 'uint256',
            },
        ],
        name: 'proveInboxContainsMessage',
        outputs: [
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256',
            },
            {
                internalType: 'bytes32',
                name: '',
                type: 'bytes32',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'sequencer',
        outputs: [
            {
                internalType: 'address',
                name: '',
                type: 'address',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'address',
                name: 'addr',
                type: 'address',
            },
            {
                internalType: 'bool',
                name: 'isSequencer',
                type: 'bool',
            },
        ],
        name: 'setIsSequencer',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'uint256',
                name: 'newMaxDelayBlocks',
                type: 'uint256',
            },
            {
                internalType: 'uint256',
                name: 'newMaxDelaySeconds',
                type: 'uint256',
            },
        ],
        name: 'setMaxDelay',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
];
class ISequencerInbox__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.ISequencerInbox__factory = ISequencerInbox__factory;
ISequencerInbox__factory.abi = _abi;
