// @typechain/truffle-v4/dist/codegen/events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.codegenAllPossibleEvents = exports.codegenEventsDeclarations = void 0;
const typechain_1 = require("typechain");
const lodash_1 = require("lodash");
const types_1 = require("./types");
function codegenEventsDeclarations(contract) {
    return lodash_1.values(contract.events)
        .filter((e) => !hasNamelessEvents(e))
        .map((e) => {
        if (e.length === 1) {
            return codegenSingleEventsDeclaration(e[0]);
        }
        else {
            return codegenOverloadEventsDeclaration(e);
        }
    })
        .join('\n');
}
exports.codegenEventsDeclarations = codegenEventsDeclarations;
function codegenAllPossibleEvents(contract) {
    const allPossibleEvents = lodash_1.values(contract.events)
        .filter((e) => !hasNamelessEvents(e))
        .map((e) => e[0])
        .filter((e) => !e.isAnonymous)
        .map((e) => e.name);
    if (allPossibleEvents.length === 0) {
        return `type AllEvents = never`;
    }
    return `type AllEvents = ${allPossibleEvents.join(' | ')};`;
}
exports.codegenAllPossibleEvents = codegenAllPossibleEvents;
function codegenOverloadEventsDeclaration(e) {
    const eventsDecls = e.map((e) => codegenSingleEventsDeclaration(e, typechain_1.getFullSignatureAsSymbolForEvent(e)));
    const union = `type ${e[0].name} = ${e.map((e) => typechain_1.getFullSignatureAsSymbolForEvent(e)).join('|')}`;
    return `
  ${eventsDecls.join('\n')}

  ${union}
  `;
}
function codegenSingleEventsDeclaration(e, overloadName) {
    return `
  export interface ${overloadName !== null && overloadName !== void 0 ? overloadName : e.name} {
    name: "${e.name}"
    args: ${codegenOutputTypesForEvents(e.inputs)}
  }
  `;
}
function codegenOutputTypesForEvents(outputs) {
    return `{
    ${outputs.map((param) => param.name + ':' + types_1.codegenOutputType(param.type)).join(', ')}
  }`;
}
function hasNamelessEvents(events) {
    return lodash_1.some(events, (e) => lodash_1.some(e.inputs, (e) => !e.name));
}
