// @typechain/truffle-v5/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const typechain_1 = require("typechain");
const ts_generator_1 = require("ts-generator");
const path_1 = require("path");
const fs_1 = require("fs");
const codegen_1 = require("./codegen");
const contracts_1 = require("./codegen/contracts");
const DEFAULT_OUT_PATH = './types/truffle-contracts/';
class Truffle extends ts_generator_1.TsGeneratorPlugin {
    constructor(ctx) {
        super(ctx);
        this.name = 'Truffle';
        this.contracts = [];
        const { cwd, rawConfig } = ctx;
        this.outDirAbs = path_1.resolve(cwd, rawConfig.outDir || DEFAULT_OUT_PATH);
    }
    transformFile(file) {
        const abi = typechain_1.extractAbi(file.contents);
        const isEmptyAbi = abi.length === 0;
        if (isEmptyAbi) {
            return;
        }
        const name = typechain_1.getFilename(file.path);
        const documentation = typechain_1.extractDocumentation(file.contents);
        const contract = typechain_1.parse(abi, name, documentation);
        this.contracts.push(contract);
        return {
            path: path_1.join(this.outDirAbs, `${contract.name}.d.ts`),
            contents: contracts_1.codegenContract(contract),
        };
    }
    afterRun() {
        return [
            {
                path: path_1.join(this.outDirAbs, 'index.d.ts'),
                contents: codegen_1.codegenArtifactHeaders(this.contracts),
            },
            {
                path: path_1.join(this.outDirAbs, 'types.d.ts'),
                contents: fs_1.readFileSync(path_1.join(__dirname, '../static/types.d.ts'), 'utf-8'),
            },
        ];
    }
}
exports.default = Truffle;
