// @codechecks/client/dist/ci-providers/GithubActions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const errors_1 = require("../utils/errors");
// github actions have explicit event for PR creation but we cant use it since it fired only once
// in future we need to build a mechanism to get pull request information using API since it's impossible to get from the inside the PR
class GithubActions {
    constructor(env) {
        this.env = env;
    }
    isCurrentlyRunning() {
        return !!this.env["GITHUB_WORKFLOW"];
    }
    getPullRequestID() {
        return undefined;
    }
    getCurrentSha() {
        const sha = this.env["GITHUB_SHA"];
        if (!sha) {
            throw errors_1.crash("Couldn't get target SHA");
        }
        return sha;
    }
    isFork() {
        // This is only set for forked repositories
        // @see https://help.github.com/en/articles/virtual-environments-for-github-actions#default-environment-variables
        return !!this.env["GITHUB_HEAD_REF"];
    }
    getProjectSlug() {
        const projectSlug = this.env["GITHUB_REPOSITORY"];
        if (!projectSlug) {
            throw errors_1.crash("Missing GITHUB_REPOSITORY");
        }
        return projectSlug;
    }
    supportsSpeculativeBranchSelection() {
        return true;
    }
}
exports.GithubActions = GithubActions;
